cpp/src/fil/fil.cu

```diff
@@ -625,13 +604,108 @@ int max_depth(const tl::ModelImpl<T, L>& model)
   return depth;
 }
 
-inline void adjust_threshold(
+// constructs a vector of size max_fid (number of features, or columns) from a Treelite tree,
+// where each feature has a maximum matching category and number of categorical nodes
+template <typename T, typename L>
+inline std::vector<int> max_matching_cat(const tl::Tree<T, L>& tree, int max_fid)
+{
+  std::vector<int> res(max_fid);
+  std::stack<int> stack;
+  stack.push(tree_root(tree));
+  while (!stack.empty()) {
+    int node_id = stack.top();
+    stack.pop();
+    while (!tree.IsLeaf(node_id)) {
+      if (tree.SplitType(node_id) == tl::SplitFeatureType::kCategorical) {
+        std::vector<uint32_t> mmv = tree.MatchingCategories(node_id);
+        int max_matching_cat;
+        if (mmv.size() > 0) {
+          // in `struct cat_feature_counters` and GPU structures, max matching category is an int
+          // cast is safe because all precise int floats fit into ints, which are asserted to be 32
+          // bits
+          max_matching_cat = mmv.back();
+          ASSERT(max_matching_cat <= MAX_PRECISE_INT_FLOAT,
+                 "FIL cannot infer on "
+                 "more than %d matching categories",
+                 MAX_PRECISE_INT_FLOAT);
+        } else {
+          max_matching_cat = -1;
+        }
+        int* max_matching_res = &res[tree.SplitIndex(node_id)];
+        *max_matching_res     = std::max(*max_matching_res, max_matching_cat);
+      }
+      stack.push(tree.LeftChild(node_id));
+      node_id = tree.RightChild(node_id);
+    }
+  }
+  return res;
+}
+
+// constructs a vector of size max_fid (number of features, or columns) from a Treelite tree,
+// where each feature has a maximum matching category and number of categorical nodes
+template <typename T, typename L>
+inline std::size_t bit_pool_size(const tl::Tree<T, L>& tree, cat_sets_owner& cat_sets)
+{
+  std::size_t size = 0;
+  std::stack<int> stack;
+  stack.push(tree_root(tree));
+  while (!stack.empty()) {
+    int node_id = stack.top();
+    stack.pop();
+    while (!tree.IsLeaf(node_id)) {
+      if (tree.SplitType(node_id) == tl::SplitFeatureType::kCategorical &&
+          tree.HasMatchingCategories(node_id)) {
+        int fid = tree.SplitIndex(node_id);
+        size += cat_sets.accessor().sizeof_mask(fid);
+        ++cat_sets.n_nodes[fid];
+      }
+      stack.push(tree.LeftChild(node_id));
+      node_id = tree.RightChild(node_id);
+    }
+  }
+  return size;
+}
+
+void vec_max(std::vector<int>& dst, const std::vector<int>& extra)
```

Comment:
Could we rename this to `elementwise_max`? I could see myself reading `vec_max` and assuming it took the max of elements in a vector.